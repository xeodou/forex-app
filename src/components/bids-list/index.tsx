import { CurrencyRate } from "lib/forex-api/client";
import { FC } from "react";

export const BidsList: FC<{ rates: CurrencyRate[] }> = ({ rates }) => {
  return (
    <div className="bids-list w-full">
      <header className="flex justify-around uppercase">
        <span>CUR</span>
        <span>Time</span>
        <span>Ask</span>
        <span>Bid</span>
      </header>
      <div className="h-60 overflow-y-auto flex flex-col">
        {rates.map((rate, index) => (
          <ul
            key={rate.time_stamp}
            className={`flex justify-around ${index % 2 === 0 ? "bg-slate-300" : "bg-slate-100 "}`}
          >
            <li>{`${rate.from}/${rate.to}`}</li>
            <li>{new Date(rate.time_stamp).toLocaleString()}</li>
            <li>{rate.ask}</li>
            <li>{rate.bid}</li>
          </ul>
        ))}
      </div>
    </div>
  );
};
