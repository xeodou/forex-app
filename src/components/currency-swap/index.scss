.currency-header {
  &--indicator {
    width: 30px;
  }
}
