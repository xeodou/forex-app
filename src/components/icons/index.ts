export * from "./icon-swap";
export * from "./icon-refresh";
export * from './icon-close';
