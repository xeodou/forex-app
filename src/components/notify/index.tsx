import { useNotify } from "hooks";
import { Item } from "./_item";

export const Notify = () => {
  const { stack, removeNotifyItem } = useNotify();

  const lastItem = stack[stack.length - 1];

  const onDismiss = (id: number) => {
    removeNotifyItem(id);
  };

  return <div className="notify-container">{lastItem && <Item item={lastItem} onDismiss={onDismiss} />}</div>;
};
