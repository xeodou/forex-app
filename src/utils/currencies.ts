export const currencies = ["USD", "JPY", "BGN", "CNY"];
