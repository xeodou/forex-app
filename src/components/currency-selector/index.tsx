import { FC } from "react";
import { currencies } from "utils/currencies";

export interface CurrencyHeaderProps {
  // On selected currency changes
  onSelected?: (currency: string) => void;
  // Specific the default selected currency
  value?: string;
}

export const CurrencySelector: FC<CurrencyHeaderProps> = ({ onSelected, value }) => {
  const onChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    onSelected && onSelected(event.target.value);
  };

  return (
    <select
      onChange={onChange}
      value={value}
      className="h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm "
    >
      {currencies.map((currency) => (
        <option key={currency} value={currency}>
          {currency}
        </option>
      ))}
    </select>
  );
};
