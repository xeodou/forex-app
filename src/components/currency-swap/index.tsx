import { CurrencySelector } from "components/currency-selector";
import { IconSwap } from "components/icons";
import { useCurrencyPair } from "hooks/use-currency-pair";

export const CurrencySwap = ({}) => {
  const [currencyPair, setCurrencyPair] = useCurrencyPair();

  const onFromCurrencyChange = (from: string) => {
    setCurrencyPair([from, currencyPair[1]]);
  };

  const onToCurrencyChange = (to: string) => {
    setCurrencyPair([currencyPair[0], to]);
  };

  const onSwitchClick = () => {
    setCurrencyPair([currencyPair[1], currencyPair[0]]);
  };

  return (
    <div className="flex items-center">
      <CurrencySelector onSelected={onFromCurrencyChange} value={currencyPair[0]} />
      <button
        type="button"
        aria-label="Switch currencies"
        className="font-medium text-sm p-2.5 text-center inline-flex items-center"
        onClick={onSwitchClick}
      >
        <IconSwap></IconSwap>
      </button>
      <CurrencySelector onSelected={onToCurrencyChange} value={currencyPair[1]} />
    </div>
  );
};
