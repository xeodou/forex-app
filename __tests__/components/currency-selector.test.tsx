import { act, fireEvent } from "@testing-library/react";
import { render, unmountComponentAtNode } from "react-dom";
import pretty from "pretty";
import { CurrencySelector } from "components/currency-selector";
import { currencies } from "utils/currencies";

describe("currency-selector", () => {
  let container: HTMLDivElement;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null as unknown as HTMLDivElement;
  });

  it("should render the selector", () => {
    act(() => {
      render(<CurrencySelector />, container);
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();

    expect(container.querySelector("select")).toBeTruthy();
    expect(container.querySelectorAll("option").length).toBe(currencies.length);
  });

  it("should render with default value", () => {
    const currency = currencies[currencies.length - 1];

    act(() => {
      render(<CurrencySelector value={currency} />, container);
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();

    expect(container.querySelector("select")?.value).toBe(currency);
  });

  it("should select different option", () => {
    const onSelected = jest.fn();

    act(() => {
      render(<CurrencySelector onSelected={onSelected} />, container);
    });

    const select = container.querySelector("select") as HTMLSelectElement;
    expect(select).toBeTruthy();

    fireEvent.change(select, { target: { value: currencies[1] } });
    expect(onSelected).toHaveBeenCalledWith(currencies[1]);

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});
