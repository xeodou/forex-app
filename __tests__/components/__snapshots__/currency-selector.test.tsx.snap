// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`currency-selector should render the selector 1`] = `
"<select class=\\"h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm \\">
  <option value=\\"USD\\">USD</option>
  <option value=\\"JPY\\">JPY</option>
  <option value=\\"BGN\\">BGN</option>
  <option value=\\"CNY\\">CNY</option>
</select>"
`;

exports[`currency-selector should render with default value 1`] = `
"<select class=\\"h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm \\">
  <option value=\\"USD\\">USD</option>
  <option value=\\"JPY\\">JPY</option>
  <option value=\\"BGN\\">BGN</option>
  <option value=\\"CNY\\">CNY</option>
</select>"
`;

exports[`currency-selector should select different option 1`] = `
"<select class=\\"h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm \\">
  <option value=\\"USD\\">USD</option>
  <option value=\\"JPY\\">JPY</option>
  <option value=\\"BGN\\">BGN</option>
  <option value=\\"CNY\\">CNY</option>
</select>"
`;
